timelock-server/src/main/java/com/palantir/atlasdb/timelock/clock/ClockSkewMonitor.java

```diff
@@ -0,0 +1,174 @@
+        this.shouldRunClockSkewMonitor = shouldRunClockSkewMonitor;
+    }
+
+    @Override
+    public void run() {
+        while (true) {
+            try {
+                sleep(sleepTimeMillis);
+                if (shouldRunClockSkewMonitor.get()) {
+                    runInternal();
+                }
+            } catch (InterruptedException e) {
+                logger.info("The clockSkewMonitor thread was interrupted. Please restart the service to restart"
+                        + " the clock");
+            }
+        }
+    }
+
+    private void runInternal() {
+        for (Map.Entry<String, ClockService> entry : monitors.entrySet()) {
+            String server = entry.getKey();
+            ClockService monitor = entry.getValue();
+
+            long localTimeAtStart = System.nanoTime();
+            long remoteSystemTime = monitor.getSystemTimeInNanos();
+            long localTimeAtEnd = System.nanoTime();
+
+            RequestTime previousRequest = previousRequests.get(server);
+            RequestTime newRequest = new RequestTime(localTimeAtStart, localTimeAtEnd, remoteSystemTime);
+            compareClockSkew(server, newRequest, previousRequest);
+        }
+    }
+
+    private void compareClockSkew(String server, RequestTime newRequest, RequestTime previousRequest) {
+        if (previousRequest == null) {
+            previousRequests.put(server, newRequest);
+            return;
+        }
+
+        long maxSkew = newRequest.localTimeAtEnd - previousRequest.localTimeAtStart;
+        long minSkew = newRequest.localTimeAtStart - previousRequest.localTimeAtEnd;
+        long remoteSkew = newRequest.remoteSystemTime - previousRequest.remoteSystemTime;
+
+        if (minSkew > maxTimeSincePreviousRequestNanos) {
+            logger.debug("It's been a long time since we last queried the server."
+                            + " Ignoring the skew, since it's not representative.",
+                    SafeArg.of("remoteSkew", remoteSkew));
+            previousRequests.put(server, newRequest);
+            return;
+        }
+
+        // maxSkew - minSkew = time for previous request and current request to complete.
+        if (maxSkew - minSkew > 2 * maxRequestTimeNanos) {
+            logger.debug("A request took too long to complete."
+                            + " Ignoring the skew, since it's not representative.",
+                    SafeArg.of("requestsDuration", maxSkew - minSkew));
+            previousRequests.put(server, newRequest);
+            return;
+        }
+
+        if (remoteSkew < minSkew) {
+            logger.info("Remote clock pace lower than expected",
+                    SafeArg.of("remoteSkew", remoteSkew), SafeArg.of("minSkew", minSkew));
+
+            metricRegistry.histogram(String.format("clock-skew-%s-below-histogram", server))
+                    .update(minSkew - remoteSkew);
+            metricRegistry.counter(String.format("clock-skew-%s-below-counter", server)).inc();
```

Comment:
Also, not sure we care about distinguishing between above and below. Could probably just use absolute value